### DISCUSSION TITLE
Clarification

### DISCUSSION THREAD

**Discussion by sitela** [grimme-lab/xtb-python#39] [Q&A]
Hello,

this is not a bug report, but rather a clarification request. I would like to use the code for an MD simulation. One thing I can't understand, though, is how I can obtain the value for the energy after atomic coordinates are changed by the algorithm that implements the dynamics. I've read all the docs but only single-point calculations are described, where atomic positions are input just once upon creation of the calculator object.

In the ASE implementation the energy was a property of the molecule object, so whenever positions were changed, the energy could be recalculated accordingly. Here, however, the calculator does not seem to depend on the molecule object. Thanks!

**Comment by awvwgk** [comment]
The details are somewhat hidden in the implementation, the `xtb` C-API (in `xtb.libxtb`) provides four objects: `environment`, `molecule`, `calculator` and `results`.

While this is a good design for C and Fortran, it does not really match the Python philosophy, therefore the Python API is implementing the wrapper classes containing all relevant API objects. A `Calculator` from `xtb.interface` contains beside the `calculator` also a `molecule` and an `environment`, this allows to bundle everything in one Python object and hide the details of the C-API. This means the Python `Calculator` can also use an update to change the cartesian coordinates.

In ASE the concept is inverted, an `Atoms` object has its (ASE) `Calculator`. Therefore, the Python `Calculator` for `xtb` is wrapped in an ASE `Calculator` and the necessary checks for cache invalidation and updating of coordinates are hidden inside from the user, as they should happen automatically as the `Atoms` is changed and new properties are requested.

So it largely depends on which layer you are looking at, C-API, Python-API or ASE-API, they focus on different concepts, abstract different tasks and target different audiences. Depending on how you want to implement the molecular dynamics a certain abstraction layer might be more suitable for your problem than the other, but the details depend on your problem.

**Reply by sitela** [reply]
Thanks for your reply. I would like to use the Python-API layer, since I think it gives the best control. I think it would be best if you could post a code snippet where the energy and gradient is calculated on a starting configuration, then a few atomic coordinates are changed, and lastly the energy and gradient are calculated on the new configuration.

Best, Sitela

**Reply by awvwgk** [reply]
The test for the external charges is almost what you are looking for:
https://github.com/grimme-lab/xtb-python/blob/fdb05a741f70d6634e086e4abbde4618e4866bdd/xtb/test_interface.py#L638

1. Start with the usual setup, defines two (very) different sets of cartesian coordinate

```python
from xtb.interface import Calculator, Param
import numpy as np
numbers = np.array([7, 1, 1, 6, 1, 1, 1])
positions1 = np.array([
    [-2.74753712457143,  0.37784000250000, -1.43187391300000],
    [-1.18171822157143, -0.72737292050000, -1.52700665300000],
    [-3.99820987557143, -0.35850981350000, -2.67576139800000],
    [-3.80581663657143,  0.22757315950000,  1.11430154300000],
    [-5.49538086757143,  1.39513306450000,  1.22999226200000],
    [-4.30486812157143, -1.67583868950000,  1.75209463100000],
    [-2.44826907957143,  0.99721383450000,  2.45438316600000],
])
positions2 = np.array([
    [ 2.99257143742857, -1.37287306250000, -1.14373597500000],
    [ 3.55243804342857, -0.85407196950000, -2.89737380800000],
    [ 4.17090944942857, -2.78580288550000, -0.62369309000000],
    [ 3.29536218742857,  0.77236389950000,  0.58285999400000],
    [ 2.78477654742857,  0.17767510450000,  2.48326881100000],
    [ 5.19909163542857,  1.56940602550000,  0.65272644000000],
    [ 1.98665062642857,  2.25726425050000,  0.02981799000000],
])
```

2. Create and evaluate the single point for this geometry

```python
calc = Calculator(Param.GFN1xTB, numbers, positions1)
res = calc.singlepoint()
```

3. Update positions of the `Molecule` object in the `Calculator` and perform another single point calculation, restarting from the previous results

```python
calc.update(positions2)
res = calc.singlepoint(res)
```

4. Repeat step 3

You still have to query the results for the energy and gradient to perform the propagation when putting step 3 into your molecular dynamics loop.

Note that the certain properties of the `Molecule` object and therefore also of the `Calculator` object are immutable once it is created. You have to delete and recreate it in case you want to change the number of atoms, the atomic numbers or its periodicity. You can of course update the cartesian coordinates.

**Reply by sitela** [reply]
Thank you, I tried the snippet and it worked. I would also like to try the xtb C-API. I've read the documentation, but unfortunately also in this case I am unsure as to how I should update the positions inside the Calculator object. Do you have a code snippet for that?

Best, Sitela

**Reply by awvwgk** [reply]
There are two ways to access the C-API of `xtb`. First, directly via C, than have a look at [the C-example in the `xtb` testsuite](https://github.com/grimme-lab/xtb/blob/master/TESTSUITE/c_api_example.c) and the header `xtb.h`. Or alternatively from Python with the CFFI objects exported from `xtb.libxtb` (the usage is similar as in C, for an example have a look at the `xtb.interface` how to pass-in numpy arrays).

**Reply by sitela** [reply]
Thank you, I've already had a look at the example, but it doesn't show how to change the coordinates inside the Calculator object and then re-calculate the energy and the gradient. Could you please show me how it can be accomplished?

**Reply by awvwgk** [reply]
Have a look at: https://github.com/grimme-lab/xtb/blob/77cf01e/include/xtb.h#L126-L131, just use the API call on the `molecule` object you want to update:
```c
xtb_updateMolecule(env, mol, positions, NULL);
if (xtb_checkEnvironment(env)) {
   xtb_showEnvironment(env, NULL);
   /* handle error & destroy objects */
}
```

**Reply by sitela** [reply]
Hi, I tried the code snippet you suggested: it compiles OK both with gcc and icc, but upon running the executable I get this error message: 

Error message is:
```
-1- xtb_api_getEnergy: Energy is not available in results.
```

My system is Ubuntu 20.10. I have just installed xtb's latest version with conda. This is the output of xtb --version:

xtb --version
```
      -----------------------------------------------------------      
     |                   =====================                   |     
     |                           x T B                           |     
     |                   =====================                   |     
     |                         S. Grimme                         |     
     |          Mulliken Center for Theoretical Chemistry        |     
     |                    University of Bonn                     |     
      -----------------------------------------------------------      

   * xtb version 6.3.3 (4ae9cef) compiled by 'conda@c0b2baabbb59' on 2020-10-14

normal termination of xtb


This is the output of gcc -v:

(base) silvio@lares:~/rmd$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/10/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 10.2.0-13ubuntu1' --with-bugurl=file:///usr/share/doc/gcc-10/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-10 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-10-JvwpWM/gcc-10-10.2.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/gcc-10-JvwpWM/gcc-10-10.2.0/debian/tmp-gcn/usr,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.2.0 (Ubuntu 10.2.0-13ubuntu1)
```


As for compilation, this is the command I issued:

```
gcc xtb-capi.c -I/opt/miniconda3/include/xtb/ -L/opt/miniconda3/lib/ -lxtb -o xtb-capi
```

I hope you can help me with this problem. Best, Sitela

**Reply by awvwgk** [reply]
Without the code snippet I can't tell for sure, but it seems like you requested a property from the results object which is not present. Make sure to actually perform a single point calculation before querying the results object.

**Reply by sitela** [reply]
I used the code snippet you pointed to on this page: https://github.com/grimme-lab/xtb/blob/77cf01e/include/xtb.h#L126-L131.

**Reply by awvwgk** [reply]
Please have a detailed look at the example.

The error printout you found is generated on purpose for the example, by requesting a property from the results object before performing the calculation:

https://github.com/grimme-lab/xtb/blob/af6e157/TESTSUITE/c_api_example.c#L47-L51

Of course you have to create a suitable example for your application first.

**Reply by sitela** [reply]
Thanks, it now works. I have one more problem, tough: the coordinates in molecule are declared as constant double of fixed length. How can my program modify them and re-pass them to xtb_updateMolecule?

**Reply by awvwgk** [reply]
For your application you will probably have a struct containing the coordinate array, if you have a 3 by n contiguous array of the coordinates in Bohr you can just pass it by reference to the `xtb_updateMolecule` procedure. In any other case you have to gather it in a 3 by n contiguous array first, convert it to Bohr and than call the API.
